metals/src/main/scala/scala/meta/internal/metals/ScalafmtConfig.scala

```diff
@@ -0,0 +1,206 @@
+package scala.meta.internal.metals
+
+import java.nio.file.FileSystems
+
+import scala.collection.JavaConverters._
+import scala.util.Try
+
+import scala.meta.io.AbsolutePath
+
+import com.typesafe.config.Config
+import com.typesafe.config.ConfigFactory
+import com.typesafe.config.ConfigParseOptions
+import com.typesafe.config.ConfigSyntax
+import com.typesafe.config.parser.ConfigDocument
+import com.typesafe.config.parser.ConfigDocumentFactory
+
+object ScalafmtConfig {
+
+  sealed trait SyntaxError
+  case object SyntaxError extends SyntaxError
+
+  case class FileOverride(
+      glob: String,
+      dialect: ScalafmtDialect
+  )
+
+  sealed trait PathMatcher {
+    def matches(path: AbsolutePath): Boolean
+  }
+  object PathMatcher {
+    final class Nio(pattern: String) extends PathMatcher {
+      private val matcher = FileSystems.getDefault().getPathMatcher(pattern)
+      def matches(path: AbsolutePath): Boolean = matcher.matches(path.toNIO)
+    }
+
+    final class Regex(regex: String) extends PathMatcher {
+      private val pattern = java.util.regex.Pattern.compile(regex)
+      def matches(path: AbsolutePath): Boolean =
+        pattern.matcher(path.toString).find()
+    }
+  }
+
+  case class ConfigValues(
+      version: Option[String],
+      runnerDialect: Option[ScalafmtDialect],
+      fileOverrides: Map[String, ScalafmtDialect],
+      includeFilters: List[PathMatcher],
+      excludeFilters: List[PathMatcher]
+  ) {
+
+    def overrideFor(path: AbsolutePath): Option[ScalafmtDialect] = {
+      val fs = FileSystems.getDefault
+      fileOverrides
+        .find { case (pm, _) =>
+          val matcher = fs.getPathMatcher(pm)
+          matcher.matches(path.toNIO)
+        }
+        .map(_._2)
+    }
+
+    def isExcluded(path: AbsolutePath): Boolean = {
+      excludeFilters.exists(_.matches(path)) && !includeFilters.exists(
+        _.matches(path)
+      )
+    }
+  }
+
+  def update(
+      configText: String,
+      version: Option[String],
+      runnerDialect: Option[ScalafmtDialect]
+  ): String = {
+
+    def docFrom(s: String): ConfigDocument = {
+      val options = ConfigParseOptions.defaults().setSyntax(ConfigSyntax.CONF)
+      ConfigDocumentFactory.parseString(s, options)
+    }
+
+    def withUpdatedVersion(content: String, v: String): String = {
+      val doc = docFrom(content)
+      if (doc.hasPath("version"))
+        doc.withValueText("version", '"' + v + '"').render
+      else {
+        // prepend to the beggining of file
+        val sb = new StringBuilder
+        sb.append(s"""version = "$v"""")
+        sb.append(System.lineSeparator)
+        sb.append(content)
+        sb.toString
+      }
+    }
+
+    def withUpdatedDialect(content: String, d: ScalafmtDialect): String = {
+      val doc = docFrom(content)
+      if (doc.hasPath("runner.dialect"))
+        doc.withValueText("runner.dialect", d.value).render
+      else {
+        // append to the end
+        val sb = new StringBuilder
```

Comment:
Could we extract that part to a common function in order no to repeat writing to config?

Reply:
There is a difference:
- for `dialect` it is `append`
- for `version` it is `prepend` (version usually goes first)

Reply:
Does it make a difference where the version is put?

Reply:
I believe that there is an unspoken rule about putting the version at the beginning of file